sdk/src/node/rpc/mod.rs

```diff
@@ -56,83 +58,110 @@ pub enum QueryMode {
     Full,
 }
 
-pub trait LedgerRpcProvider {
-    type SlotResponse: Serialize;
-    type BatchResponse: Serialize;
-    type TxResponse: Serialize;
-    type EventResponse: Serialize;
+#[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
+pub struct SlotResponse<B, Tx> {
+    pub number: u64,
+    pub hash: [u8; 32],
+    pub batch_range: std::ops::Range<u64>,
+    pub batches: Option<Vec<ItemOrHash<BatchResponse<B, Tx>>>>,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
+pub struct BatchResponse<B, Tx> {
+    pub hash: [u8; 32],
+    pub tx_range: std::ops::Range<u64>,
+    pub txs: Option<Vec<ItemOrHash<TxResponse<Tx>>>>,
+    pub custom_receipt: B,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
+pub struct TxResponse<Tx> {
+    pub hash: [u8; 32],
+    pub event_range: std::ops::Range<u64>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub body: Option<Vec<u8>>,
+    #[serde(flatten)]
+    pub custom_receipt: Tx,
+}
 
-    fn get_head(&self) -> Result<Option<Self::SlotResponse>, anyhow::Error>;
+#[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum ItemOrHash<T> {
+    Hash([u8; 32]),
+    Full(T),
+}
 
-    fn get_slots(
+pub trait LedgerRpcProvider {
```

Comment:
This is a good question. I think (surprisingly), that adding the generics to the trait actually makes it more difficult to use! 

This is because the implementer of this trait (LedgerDB) doesn't have generics of its own - so it has to have a blanket impl for all choices of `B` and `T`. Which means that every time you want to use the trait, you still have to provide the generics, and you have to do it using the ugly fully-qualified syntax `< LedgerRpcProvider<MyBatchType, MyTxType>>::get_head(ledger_db)`.